### PR TITLE
Add VRF proof verification / generation to the Schnorrkel wrapper

### PR DESCRIPTION
**Pull request by David-Petrov**

### CODE REVIEW

**Comment by David-Petrov** [comment]
[KNOWN ISSUE] Build automation is broken. The original `jar` task for managing the native rust lib (removed by [015023e](https://github.com/LimeChain/polkaj/pull/2/commits/015023eec6e683886c6fda2596186904461c3f4c)) was meant to pick up the lib as output from cargo (built for the current architecture) and place it in its corresponding place. This, however, is a single-target automation.
Our fork of PolkaJ requires multi-target builds, so this approach is not suitable, thus `jar` has been removed. 
For now, manual intervention is necessary: paste the rust lib manually from `./polkaj-schnorrkel/build/rust/release` (that's the cargo build result; path relative from repository root) into its corresponding target architecture folder under `./polkaj-schnorrkel/src/main/resources/native`, before then building again to obtain the properly updated `.jar`.

Unfortunately, compilation for all target architectures must happen manually in a similar way as described above, before officially "deploying a new version" for multi-arch usage.

Is it worth investing effort into polishing the automation...?

[RESOLVED]: Discussed with @georg-getz, polishing the automation is not worth it for now, as priorities are different. This goes into `TODO.md`. For now, a detailed description of this necessary manual intervention is to be added to this fork's `README.md`.

File: polkaj-schnorrkel/src/rust/src/merlin_jni/mod.rs
Comment (David-Petrov) [inline polkaj-schnorrkel/src/rust/src/merlin_jni/mod.rs]: That's a patch fix. I mean the field being a `Vec<Box<[u8]>>`, which corresponds to an `ArrayList<byte[]>` instead of simply a `Box<[u8]>`, which corresponds to a `byte[]`. Some weird, deeper rooted trait-resolution issues arise, which (I suspect) is a bug with robusta itself... Possible solutions:
1. Fix robusta (I've tried, but not to much avail).
2. Throw robusta away and couple ourselves to primitive `jni` signatures, thus either manually mimick the `TryFrom/IntoJavaValue` traits for more complex objects or simply expand signatures to unpack the inner state representations (and end up with a lot of function arguments...). The former will require work, the latter is lazy but scales really badly.
3. Leave robusta in for its good parts - mainly the `TryFrom/IntoJavaValue` traits used for now - but don't try to derive them and instead manually implement them (I'm not yet sure how feasible this is).

This issue also directly affects the decision about the (for now) ugly return strategy on the `_vrfSign` native function.

File: polkaj-schnorrkel/src/main/java/io/emeraldpay/polkaj/schnorrkel/VrfOutputAndProof.java
Comment (georg-getz) [inline polkaj-schnorrkel/src/main/java/io/emeraldpay/polkaj/schnorrkel/VrfOutputAndProof.java:39]: Remove and add @Getter annotation.

File: polkaj-schnorrkel/src/main/java/io/emeraldpay/polkaj/schnorrkel/VrfOutputAndProof.java
Comment (georg-getz) [inline polkaj-schnorrkel/src/main/java/io/emeraldpay/polkaj/schnorrkel/VrfOutputAndProof.java]: add `final`

File: polkaj-schnorrkel/src/main/java/io/emeraldpay/polkaj/schnorrkel/VrfOutputAndProof.java
Comment (David-Petrov) [reply polkaj-schnorrkel/src/main/java/io/emeraldpay/polkaj/schnorrkel/VrfOutputAndProof.java:39]: No dependency (e.g. Lombok) is currently included that provides this, not worth it IMO for now.

File: polkaj-schnorrkel/src/main/java/io/emeraldpay/polkaj/merlin/TranscriptData.java
Comment (David-Petrov) [inline polkaj-schnorrkel/src/main/java/io/emeraldpay/polkaj/merlin/TranscriptData.java]: Self explanatory. Still, leaving this convenience method in the public interface of this here `TranscriptData` (inspired by the original rust `schnorrkel`) might not be such a good idea in the Java world, since we don't have an unified `u64` type, so now it's up to the caller to understand my `NOTE:` in the javadoc... which stinks.

**Review by georg-getz — APPROVED** [review]

File: polkaj-schnorrkel/src/rust/src/merlin_jni/mod.rs
Comment (David-Petrov) [inline polkaj-schnorrkel/src/rust/src/merlin_jni/mod.rs]: That's also a topic I'm rather unsure about. Leaking the memory (effectively telling rust "leave it up to me to handle") should be okay since the memory is actually owned and managed by the Java runtime. I am, however, lacking the deepest understanding as of what possible issues we could have with that type of memory management :D

File: polkaj-schnorrkel/src/main/resources/native/macos/arm/libpolkaj_schnorrkel.dylib
Comment (David-Petrov) [inline polkaj-schnorrkel/src/main/resources/native/macos/arm/libpolkaj_schnorrkel.dylib:1]: The removal of this file has to do with forcing recompilation of the native rust lib, since automatic rehydration was broken. Will have to be manually added again (with the finalised changes on the rust side) directly before merging this PR... Or we fix the automations.

File: polkaj-schnorrkel/src/rust/src/merlin_jni/mod.rs
Comment (David-Petrov) [reply polkaj-schnorrkel/src/rust/src/merlin_jni/mod.rs]: After considerable headaches, all three approaches turn out to be a bigger headache than expected. Thus, for now the hacks remain as-is (I deemed them not that big of a pain), an attempt at a proper documentation of the decisions made has been made.

Our biggest hope lies with the active participation in upgrading robusta to jni to 0.21, so we're following those PRs: https://github.com/giovanniberti/robusta/pull/50 (and https://github.com/giovanniberti/robusta/pull/59 related).

May need later revisiting.

File: polkaj-schnorrkel/src/main/java/io/emeraldpay/polkaj/merlin/TranscriptData.java
Comment (David-Petrov) [reply polkaj-schnorrkel/src/main/java/io/emeraldpay/polkaj/merlin/TranscriptData.java]: I've removed it. We'll manage this u64 representation as a `byte[]` from Fruzhin.

File: polkaj-schnorrkel/src/rust/src/merlin_jni/mod.rs
Comment (David-Petrov) [reply polkaj-schnorrkel/src/rust/src/merlin_jni/mod.rs]: After digging into this for a little bit, I managed to convince myself that's rather OK. Proper documentation has been added on the intentionality of this piece of code.

File: polkaj-schnorrkel/src/main/resources/native/macos/arm/libpolkaj_schnorrkel.dylib
Comment (David-Petrov) [reply polkaj-schnorrkel/src/main/resources/native/macos/arm/libpolkaj_schnorrkel.dylib:1]: All has been manually rebuilt.

**Pull request merged**
